crates/noirc_frontend/src/parser/grammar.bnf

```diff
@@ -0,0 +1,227 @@
+
+/// expression_with_precedence(Highest, ExprParser, true) ::= type_expression_term
+/// expression_with_precedence(Highest, ExprParser, false) ::= term
+/// expression_with_precedence(precedence, ExprParser, true) ::= expression_with_precedence(precedence.next_type_precedence(), ExprParser, true)
+///                                                              (
+///                                                                 operator_with_precedence(precedence)
+///                                                                 expression_with_precedence(precedence.next_type_precedence(), ExprParser, true)
+///                                                              )*
+/// expression_with_precedence(precedence, ExprParser, false) ::= expression_with_precedence(precedence.next(), ExprParser, false)
+///                                                              (
+///                                                                 operator_with_precedence(precedence)
+///                                                                 expression_with_precedence(precedence.next(), ExprParser, false)
+///                                                              )*
+
+/// EQUIVALENT TO:
+
+expression ::= eq_expr
+eq_expr ::= eq_expr ('=' | '!=') or_expr | or_expr
+or_expr ::= or_expr '|' xor_expr | xor_expr
+xor_expr ::= xor_expr '^' and_expr | and_expr
+and_expr ::= and_expr '&' less_greater_expr | less_greater_expr
+less_greater_expr :;= less_greater_expr ('<' | '<=' | '>' | '>=') shift_expr
+shift_expr ::= shift_expr ('<<' | '>>') sum_expr | sum_expr
+sum_expr ::= sum_expr ('+' | '-') product_expr | product_expr
+product_expr ::= product_expr ('/' | '*') term | term
+
+type_expression ::= type_eq_expr
+type_eq_expr ::= type_eq_expr ('=' | '!=') type_sum_expr | type_sum_expr
+type_sum_expr ::= type_sum_expr ('+' | '-') type_product_expr | type_product_expr
+type_product_expr ::= type_product_expr ('/' | '*') type_expression_term | type_expression_term
+
+////////////////////////////////////////////////////////////////////////////////
+
+term ::= not | negation | atom_or_right_unary
+
+atom_or_right_unary ::= atom (call_rhs | array_rhs | cast_rhs | member_rhs)*
+call_rhs ::= '(' expression_list ')'
+array_rhs ::= '[' expression ']'
+cast_rhs ::= 'as' type
+member_rhs ::= '.' field_name ('(' expression_list ')')?
+
+if_expr ::= 'if' expression block_expr ('else' (block_expr | if_expr))? 
+
+lambda ::= '|' lambda_parameters '|' lambda_return_type expression
+
+for_expr ::= 'for' ident 'in' for_range block_expr 
+
+for_range ::= expression '..' expression | expression
+
+array_expr ::= standard_array | array_sugar
+
+standard_array ::= '[' expression ']'
+
+array_sugar ::= '[' expression ';' expression ']' 
+
+expression_list ::= (expression%','+ ','?)?
+
+not ::= '!' term 
+
+negation ::= '-' term
+
+atom ::= '(' expression ')' | tuple | if_expr | for_expr | array_expr | constructor | lambda | block | variable | literal
```

Comment:
Unnecessary code: `tuple` covers `'(' expression ')'`

a "(foo)" can be parsed either as '(' expression ')' or a single-element tuple

this means the '(' expression ')' part could be removed